debezium-connector-sqlserver/src/main/java/io/debezium/connector/sqlserver/SqlServerStreamingChangeEventSource.java

```diff
@@ -83,98 +99,186 @@ public void execute(ChangeEventSourceContext context) throws InterruptedExceptio
+
+                        final int tableCount = resultSets.length;
+                        final ChangeTablePointer[] changeTables = new ChangeTablePointer[tableCount];
+                        final ChangeTable[] tables = tablesSlot.get();
 
-                    for (;;) {
-                        ChangeTable tableSmallestLsn = null;
                         for (int i = 0; i < tableCount; i++) {
-                            final ChangeTable changeTable = changeTables[i];
-                            if (changeTable.isCompleted()) {
+                            changeTables[i] = new ChangeTablePointer(tables[i], resultSets[i]);
+                            changeTables[i].next();
+                        }
+
+                        for (;;) {
+                            ChangeTablePointer tableSmallestLsn = null;
+                            for (ChangeTablePointer changeTable: changeTables) {
+                                if (changeTable.isCompleted()) {
+                                    continue;
+                                }
+                                if (tableSmallestLsn == null || changeTable.compareTo(tableSmallestLsn) < 0) {
+                                    tableSmallestLsn = changeTable;
+                                }
+                            }
+                            if (tableSmallestLsn == null) {
+                                // No more LSNs available
+                                break;
+                            }
+
+                            if (tableSmallestLsn.getRowLsn().compareTo(lastProcessedLsnOnStart) <= 0) {
+                                LOGGER.info("Skipping change {} as its LSN is smaller than the last recorded LSN {}", tableSmallestLsn, lastProcessedLsnOnStart);
+                                tableSmallestLsn.next();
                                 continue;
                             }
-                            if (tableSmallestLsn == null || changeTable.compareTo(tableSmallestLsn) < 0) {
-                                tableSmallestLsn = changeTable;
+                            if (tableSmallestLsn.getChangeTable().getStopLsn().isAvailable() &&
+                                    tableSmallestLsn.getChangeTable().getStopLsn().compareTo(tableSmallestLsn.getRowLsn()) <= 0) {
+                                LOGGER.debug("Skipping table change {} as its stop LSN is smaller than the last recorded LSN {}", tableSmallestLsn, tableSmallestLsn.getRowLsn());
+                                tableSmallestLsn.next();
+                                continue;
                             }
-                        }
-                        if (tableSmallestLsn == null) {
-                            // No more LSNs available
-                            break;
-                        }
+                            LOGGER.trace("Processing change {}", tableSmallestLsn);
+                            if (!schemaChangeCheckpoints.isEmpty()) {
```

Comment:
I built your changes and I ran them locally on Kafka Connect.

Test steps:
1) Table created, cdc enabled.
2) Add a record with the old schema.
3) Alter table - add column. Enable a new capture instance.
4) Add a record with the new schema.

Result:
 + The first event from step (2) looks as expected.
 - Schema change event has not been emitted.
 - the second event from step (4) has the old schema (field `height` missing).

```
CREATE TABLE products (
  id INTEGER NOT NULL AUTO_INCREMENT PRIMARY KEY,
  name VARCHAR(255) NOT NULL,
  description VARCHAR(512),
  weight FLOAT
);
ALTER TABLE products AUTO_INCREMENT = 101;
EXEC sys.sp_cdc_enable_table @source_schema = 'dbo', @source_name = 'products', @role_name = NULL, @supports_net_changes = 0;
```

```
1> insert into products (name, description, weight) values ('book', 'a book', 0.6);
2> GO
1> alter table products add height int default null;
2> GO
1> EXEC sys.sp_cdc_enable_table @source_schema = 'dbo', @source_name = 'products', @capture_instance = 'dbo_products_v2', @role_name = NULL, @supports_net_changes = 0;
2> GO
1> insert into products (name, description, weight, height) values ('book', 'a book', 0.77, 1);
2> GO
```

```
connect_1    | 2018-10-10 14:17:11,922 INFO   ||  Processing change ChangeTable [changeTable=ChangeTable [captureInstance=dbo_products, sourceTableId=inventory.dbo.products, changeTableId=inventory.cdc.dbo_products_CT, startLsn=00000023:00000b80:0043, changeTableObjectId=1269579561, stopLsn=NULL], resultSet=SQLServerResultSet:74, completed=false]
connect_1    | 2018-10-10 14:17:54,854 INFO   ||  Multiple capture instances ChangeTable [captureInstance=dbo_products, sourceTableId=inventory.dbo.products, changeTableId=inventory.cdc.dbo_products_CT, startLsn=00000023:00000b80:0043, changeTableObjectId=1269579561, stopLsn=00000026:00000018:0048] and ChangeTable [captureInstance=dbo_products, sourceTableId=inventory.dbo.products, changeTableId=inventory.cdc.dbo_products_CT, startLsn=00000023:00000b80:0043, changeTableObjectId=1269579561, stopLsn=00000026:00000018:0048] present for the same table
connect_1    | 2018-10-10 14:17:54,855 INFO   ||  Schema will be changed for ChangeTable [captureInstance=dbo_products, sourceTableId=inventory.dbo.products, changeTableId=inventory.cdc.dbo_products_CT, startLsn=00000023:00000b80:0043, changeTableObjectId=1269579561, stopLsn=00000026:00000018:0048]
connect_1    | 2018-10-10 14:18:10,858 TRACE  ||  Processing change ChangeTable [changeTable=ChangeTable [captureInstance=dbo_products_v2, sourceTableId=inventory.dbo.products, changeTableId=inventory.cdc.dbo_products_v2_CT, startLsn=00000026:00000018:0048, changeTableObjectId=1765581328, stopLsn=NULL], resultSet=SQLServerResultSet:152, completed=false]
connect_1    | 2018-10-10 14:18:10,859 DEBUG  ||  Skipping table change ChangeTable [changeTable=ChangeTable [captureInstance=dbo_products, sourceTableId=inventory.dbo.products, changeTableId=inventory.cdc.dbo_products_CT, startLsn=00000023:00000b80:0043, changeTableObjectId=1269579561, stopLsn=00000026:00000018:0048], resultSet=SQLServerResultSet:153, completed=false] as its stop LSN is smaller than the last recorded LSN 00000026:00000828:0003
```

```
{"schema":{"type":"struct","fields":[{"type":"int32","optional":false,"field":"id"}],"optional":false,"name":"mssql_server.inventory.dbo.products.Key"},"payload":{"id":110}}	{"schema":{"type":"struct","fields":[{"type":"int32","optional":false,"field":"id"},{"type":"string","optional":false,"field":"name"},{"type":"string","optional":true,"field":"description"},{"type":"double","optional":true,"field":"weight"}],"optional":true,"name":"mssql_server.inventory.dbo.products.Value"},"payload":{"id":110,"name":"book","description":"a book","weight":0.6}}
{"schema":{"type":"struct","fields":[{"type":"int32","optional":false,"field":"id"}],"optional":false,"name":"mssql_server.inventory.dbo.products.Key"},"payload":{"id":111}}	{"schema":{"type":"struct","fields":[{"type":"int32","optional":false,"field":"id"},{"type":"string","optional":false,"field":"name"},{"type":"string","optional":true,"field":"description"},{"type":"double","optional":true,"field":"weight"}],"optional":true,"name":"mssql_server.inventory.dbo.products.Value"},"payload":{"id":111,"name":"book","description":"a book","weight":0.77}}
```

Reply:
Could you please try changing the operator here https://github.com/debezium/debezium-incubator/pull/33/files#diff-041956c14f1eae2fc358015c512bb446R153 to `>=`?

Reply:
No, the problem is the fact that `schemaChangeCheckpoints` is empty. Apparenty, it is cleared in wrong place. I added an additional log just before clearing the queue and I received:

```
connect_1    | 2018-10-11 13:24:44,494 INFO   ||  Multiple capture instances ChangeTable [captureInstance=dbo_products, sourceTableId=inventory.dbo.products, changeTableId=inventory.cdc.dbo_products_CT, startLsn=00000023:00000b78:0043, changeTableObjectId=1269579561, stopLsn=00000026:000000f0:0048] and ChangeTable [captureInstance=dbo_products, sourceTableId=inventory.dbo.products, changeTableId=inventory.cdc.dbo_products_CT, startLsn=00000023:00000b78:0043, changeTableObjectId=1269579561, stopLsn=00000026:000000f0:0048] present for the same table   [io.debezium.connector.sqlserver.SqlServerStreamingChangeEventSource]
connect_1    | 2018-10-11 13:24:44,494 INFO   ||  Schema will be changed for ChangeTable [captureInstance=dbo_products, sourceTableId=inventory.dbo.products, changeTableId=inventory.cdc.dbo_products_CT, startLsn=00000023:00000b78:0043, changeTableObjectId=1269579561, stopLsn=00000026:000000f0:0048]   [io.debezium.connector.sqlserver.SqlServerStreamingChangeEventSource]
connect_1    | 2018-10-11 13:25:04,462 INFO   ||  Clearing checkpoints.   [io.debezium.connector.sqlserver.SqlServerStreamingChangeEventSource]
connect_1    | 2018-10-11 13:25:04,487 INFO   ||  Processing change ChangeTable [changeTable=ChangeTable [captureInstance=dbo_products_v2, sourceTableId=inventory.dbo.products, changeTableId=inventory.cdc.dbo_products_v2_CT, startLsn=00000026:000000f0:0048, changeTableObjectId=1765581328, stopLsn=NULL], resultSet=SQLServerResultSet:244, completed=false]   [io.debezium.connector.sqlserver.SqlServerStreamingChangeEventSource]
```

Reply:
Oh, probably the case when in one iteration there was a DDL statement but no DML statements is not covered properly.

```
if (tableSmallestLsn == null) {
    // No more LSNs available
    break;
}
```
This prevents from calling `dispatcher.dispatchSchemaChangeEvent` in this case.

Reply:
@grzegorz8 The analysis was spot on, thanks! I've fixed the issue and added a test for it. Also the source table is now used as schema source so null/not-null works correctly.

Reply:
Thanks, it works now.
However, there is no test which checks whether schema change event is generated or not.

Reply:
@jpechane Have you added such test since then?

Reply:
@gunnarmorling See `io.debezium.connector.sqlserver.SqlServerChangeTableSetIT.addColumnToTableEndOfBatch()`